### PR TITLE
Block Virtual Threads for JVMTI Inspection

### PR DESCRIPTION
**Pull request by babsingh**
- Return the corresponding `J9VMThread` for a virtual thread from
  `getVMThread`.
- Prevent the virtual thread from unmounting using the `inspectorCount`
  approach.
- The virtual thread is stopped just before `J9VMThread->threadObject` is set
  from the virtual thread to the carrier thread.
- This also prevents a virtual thread from freeing its native resources while
  it is being inspected.
- For a virtual thread, even the underlying `J9VMThread` is prevented from
  freeing its native resources using the `inspectorCount` approach.
- `isSameOrSuperClassOf` is unable to handle `NULL` from the `*_OR_NULL` 
   version. Instead, use `J9VMJAVALANGBASEVIRTUALTHREAD` in
  `IS_VIRTUAL_THREAD`.
- Moved code from `notifyJvmtiUnmountEnd` to `notifyJvmtiUnmountBegin` for blocking
before `unmount`. Also, `VirtualThread.carrierThread` is set to NULL in `unmount`.
So, removing the VirtualThread from the linked list before the last `unmount`
in `notifyJvmtiUnmountBegin` seems correct.
- Block mount after yield if a virtual thread is being inspected.
- In `getVMThread`:
  - `targetThread` will be NULL for a yielded virtual thread.
  - `JVMTI_ERROR_NONE` will be returned for a yielded virtual thread.
  - `virtualThreadInspectorCount` will be incremented for a yielded
     virtual thread.
- Reload references after the wait since they can become outdated if a GC
  occurs during the wait.
- Release VM access before the wait and re-acquire VM access after the wait.
- Decrement `virtualThreadInspectorCount` for yielded virtual threads i.e.
  `(targetThread == NULL)`.

Related: #15183. This impacts all JVMTI functions that rely on `getVMThread` and
`releaseVMThread`.

Signed-off-by: Babneet Singh <sbabneet@ca.ibm.com>

### CODE REVIEW

**Comment by babsingh** [comment]
@gacholio When a virtual thread is being inspected in JVMTI, do we need to protect the virtual thread and corresponding carrier thread from freeing their native resources?

Existing code to prevent freeing of J9VMThread native resources:
https://github.com/eclipse-openj9/openj9/blob/fc5ae4e6892b7ad809270ecb981606f4095f62da/runtime/jvmti/jvmtiHelpers.c#L128-L130
https://github.com/eclipse-openj9/openj9/blob/6a2aac87ec442d1d96863265761996c9764b19f3/runtime/vm/jvmfree.c#L200-L205

Similar code will be needed to prevent freeing of virtual thread resources?

File: runtime/jvmti/jvmtiHelpers.c
Comment (gacholio) [inline runtime/jvmti/jvmtiHelpers.c:158]: Do not use ++ here, just add 1.

Comment (gacholio) [reply runtime/jvmti/jvmtiHelpers.c:158]: Remember it's always illegal to use side-effect expressions in macros.

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

**Comment by gacholio** [comment]
Is the unmount native called before the vthread ends and when it yields? Both of these need to block if the vthread is being inspected.

Also wondering if `haltThreadForInspection` will also need to be modified in a similar fashion.

**Comment by babsingh** [comment]
> Is the unmount native called before the vthread ends and when it yields? Both of these need to block if the vthread is being inspected.

Yes, `notifyJvmtiUnmountBegin` is called before the vthread ends and when it yields. For the yield case, the boolean input parameter is false. Updated `notifyJvmtiUnmountBegin` to `block` in both cases.

Ref:
- Terminate: [VirtualThread.java#L305](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L305)
- Yield: [VirtualThread.java#L367](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L367)

**Comment by babsingh** [comment]
> Also wondering if haltThreadForInspection will also need to be modified in a similar fashion.

`haltThreadForInspection` is always called after `getVMThread`. So, `haltThreadForInspection` should work with virtual threads as-is because both the virtual thread and corresponding carrier thread (J9VMThread) are simultaneously blocked/pinned in `getVMThread`. Also, halting the carrier thread should automatically halt the mounted virtual thread. @tajila @fengxue-IS Comment if you disagree.

**Comment by gacholio** [comment]
The native should be called before the yield (so that the vmthread stack still represents the continuation, not the carrier).

**Comment by babsingh** [comment]
> The native should be called before the yield (so that the vmthread stack still represents the continuation, not the carrier).

This is true. See [VirtualThread.java#L367](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L367). The native (`notifyJvmtiUnmountBegin`) is called before `Continuation.yield`.

**Comment by gacholio** [comment]
haltThreadForInspection is called in many places outside of JVMTI, but we can ignore it for now. If a thread is halted, it will be unable to change the mount state.

**Comment by babsingh** [comment]
@EricYangIBM Can you review these changes? Also, let me know if they work correctly with your TLS changes and related tests.

File: runtime/jvmti/jvmtiHelpers.c
Comment (EricYangIBM) [inline runtime/jvmti/jvmtiHelpers.c:140]: If the virtual thread is yielded, will the carrier thread be null and return not alive?
Also `NULL == targetThread`

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:140]: Yes. See [VirtualThread.java#L367-L370](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L367-L370). In [unmount](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/3b112d4e1c05e85584248cda15d4c3c3c4c1d6be/src/java.base/share/classes/java/lang/VirtualThread.java#L345), `VirtualThread.carrierThread` is set to `NULL` during yield. So, the `targetThread` will be `NULL`. Will fix `NULL == targetThread`. Also, a `NULL` check is missing for `VirtualThread.carrierThread`.

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

Comment (EricYangIBM) [reply runtime/jvmti/jvmtiHelpers.c]: If it's yielded, the targetThread will remain null and return "not alive" even though yielded is considered alive. Could it set it to currentThread if yielded? I need it for the tls get/set.

Comment (EricYangIBM) [reply runtime/jvmti/jvmtiHelpers.c]: Also do the jvmti functions need to have the carrier J9VMThread to work for virtual threads? Can we return vmThread = currentThread always?

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: > Also do the jvmti functions need to have the carrier J9VMThread to work for virtual threads? Can we return vmThread = currentThread always?

It depends. Most of the time, the virtual thread's information will be derived from the carrier thread (J9VMThread) during inspection. The JVMTI functions won't behave correctly with `vmThread = currentThread`. For information which does not depend on the carrier thread, `vmThread = currentThread` can be used.

> If it's yielded, the targetThread will remain null and return "not alive" even though yielded is considered alive. Could it set it to currentThread if yielded? I need it for the tls get/set.

Some locations will break with `targetThread = currentThread (from getVMThread)` if a virtual thread is yielded. For tls get/set, we can make an exception for virtual threads if it is yielded. In general, we will have to evaluate each JVMTI function individually for such needs.

Comment (EricYangIBM) [reply runtime/jvmti/jvmtiHelpers.c]: Maybe we can return a different error if a virtual thread is yielded? Then handle accordingly based on the jvmti function.
Also this means that the `SetLocal(Object|Int|Long|Float|Double) ...` functions will not be fixed by this PR

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: i can return the value of `isVirtual` from `getVMThread` via an input parameter; return `JVMTI_ERROR_NONE`; but, `targetThread` will be set to NULL. if `isVirtual && (JVMTI_ERROR_NONE == rc) && (NULL == targetThread)`, we can use the `currentThread` for evaluating the virtual thread detail. `currentThread` is derived in most cases.  for `SetLocal*`, the yielded virtual thread will need to be stack walked on the current thread. ++ @gacholio @fengxue-IS @tajila

File: runtime/jcl/common/thread.cpp
Comment (tajila) [inline runtime/jcl/common/thread.cpp:577]: you can `pushObjectInSpecialFrame` then `releaseVMAccess` before the wait and do the opposite after the wait

Comment (tajila) [reply runtime/jcl/common/thread.cpp:577]: that being said, Im not sure this is the correct approach. I think you can use haltThreadForInspection if the vthread is mounted, and introduce an equivalent for unmounted threads on j9vmcontinuations. THat would be a more general approach, and I think that would benefit us because we will need something like Thread::getStateImpl and Continuation::getStackTrace

Comment (tajila) [reply runtime/jcl/common/thread.cpp:577]: perhaps you can just pin thread, and add a check in continuationEnter, that if its pinned it can't enter

Comment (babsingh) [reply runtime/jcl/common/thread.cpp:579]: - It is identical to the blocking for j.l.Thread:
https://github.com/eclipse-openj9/openj9/blob/984612f68ad11944e7ce1ecec7b57cb9a7f6eaee/runtime/vm/jvmfree.c#L200-L205

- A running virtual thread is not halted with these changes. It still runs. We only block it from unmounting similar to the above changes.
- We have to do the same for a yielded thread. We only block it from being mounted/running again until JVMTI finishes inspecting it in the yielded state. 
- Virtual thread invokes `Continuation::getStackTrace` in an unmounted state after setting the `SUSPENDED` flag : [VirtualThread.java#L862-L865](https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/703f7587b12c993e641847e9d0c5402497ef9afb/src/java.base/share/classes/java/lang/VirtualThread.java#L862-L865). The `SUSPENDED` flag prevents it to be rescheduled. So, no blocking or synchronization is needed for `Continuation::getStackTrace`.

Comment (babsingh) [reply runtime/jcl/common/thread.cpp:579]: > you can pushObjectInSpecialFrame then releaseVMAccess before the wait and do the opposite after the wait

Updated.

File: runtime/jvmti/jvmtiHelpers.c
Comment (tajila) [inline runtime/jvmti/jvmtiHelpers.c:175]: does isVritual need to be a parameter?

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: There are few cases where the caller doesnt care. If you pass null to `getVMThread` then what do you pass into `releaseVMThread`

File: runtime/jcl/common/thread.cpp
Comment (tajila) [reply runtime/jcl/common/thread.cpp:579]: Okay, perhaps this is the best way. I was hoping there was a way to reuse the pinning mechanism for this, but I don't think there is.

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:175]: > does isVritual need to be a parameter?

No. But, `IS_VIRTUAL_THREAD` will be called more than twice. In some cases, `isVirtual` will also be used in the core body of the JVMTI function (e.g. [GetThreadInfo](https://github.com/eclipse-openj9/openj9/issues/15183#issuecomment-1187580033)). This approach shares the result of `IS_VIRTUAL_THREAD` after evaluating it once in `getVMThread`.

> There are few cases where the caller doesnt care. If you pass null to getVMThread then what do you pass into releaseVMThread

Those cases will only handle platform threads from `J9VMThread->carrierThreadObject`. Virtual threads are not expected. `getVMThread` will be given `NULL`, and `releaseVMThread` will be given `FALSE`.

File: runtime/jcl/common/thread.cpp
Comment (hangshao0) [inline runtime/jcl/common/thread.cpp:627]: Is there a reason to add the `omrthread_monitor_notify_all()` call here ?

Comment (babsingh) [reply runtime/jcl/common/thread.cpp:627]: More than one thread can be blocked on the `liveVirtualThreadListMutex`. This notify will make sure all blocked threads are incrementally woken up.

Comment (hangshao0) [reply runtime/jcl/common/thread.cpp:627]: They seems to be blocked on this mutex waiting for `virtualThreadInspectorCountOffset` to be 0. I don't see `virtualThreadInspectorCountOffset` changed in this function.

Comment (babsingh) [reply runtime/jcl/common/thread.cpp:627]: It is incremented in `getVMThread` and decremented in `releaseVMThread`. I would suggest viewing the existing `J9VMThread->inspectorCount` changes. The new changes are a replica of them.

Comment (hangshao0) [reply runtime/jcl/common/thread.cpp:627]: I understand why releaseVMThread() called omrthread_monitor_notify_all(vm->liveVirtualThreadListMutex). It changes the inspector count and it is inside if (0 == vthreadInspectorCount) {

There was a typo in my previous comment. By virtualThreadInspectorCountOffset I mean virtualThreadInspectorCount

File: runtime/jvmti/jvmtiHelpers.c
Comment (hangshao0) [inline runtime/jvmti/jvmtiHelpers.c:181]: You can move this line into line 178, `threadObject` is needed only when `NULL != thread && isVirtual`. In the `FALSE == isVirtual` case, we do not need to execute this line: `threadObject = J9_JNI_UNWRAP_REFERENCE(thread);`

File: runtime/jcl/common/thread.cpp
Comment (hangshao0) [inline runtime/jcl/common/thread.cpp:606]: The comment says `do not allow it to die`, I guess it is `do not let it to unmount`.

File: runtime/jvmti/jvmtiHelpers.c
Comment (hangshao0) [inline runtime/jvmti/jvmtiHelpers.c:188]: In the line above the new inspector count is set to `vthreadInspectorCount - 1,` shouldn't this check the new inspector count ? i.e. `(0 == (vthreadInspectorCount -1))`

File: runtime/jcl/common/thread.cpp
Comment (babsingh) [reply runtime/jcl/common/thread.cpp:627]: After our discussion during lunch, I'm convinced that this notify is not needed.

File: runtime/jvmti/jvmtiHelpers.c
Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: In `jvmtiGetOSThreadID` for example. The `isVirtual` variable is created but is never actaully needed. Is it possible for `releaseVMThread` to fetch the data from `thread` ?

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

File: runtime/jcl/common/thread.cpp
Comment (babsingh) [reply runtime/jcl/common/thread.cpp]: Updated.

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

File: runtime/jcl/common/thread.cpp
Comment (babsingh) [reply runtime/jcl/common/thread.cpp]: Updated.

File: runtime/jvmti/jvmtiHelpers.c
Comment (hangshao0) [inline runtime/jvmti/jvmtiHelpers.c:185]: I suggest to add an assertion that the old inspector count is not 0 before it is decremented here. And add an assertion that the new inspector count is not 0 after incremented in `getVMThread()`.

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:175]: Such scenarios need to be skipped for yielded virtual threads `(targetThread == NULL)`. `jvmtiInterruptThread` also falls in this category. These functions

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: by `thread`, I mean `jthread thread` which is just an object.

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: what are the conditions where the threadObject can be null in `releaseVMThread`

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: I guess what Im saying is:
if `thread` exists, then you can also just unwrap it and determine if its a virtual thread and do the necesarry parts for `vthreadInspectorCount`

If it doesnt exist, then presumably it wasn't a virtual thread or at the very least you wont need to do anything for `vthreadInspectorCount` since the counters are on the object itself

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:175]: if a user sends a NULL `thread` (as input) and `getVMThread` allows a `null`. in such a case, `isVirtual` will stay `FALSE`, and none of the new code will be invoked.

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:175]: > I dont understand this comment

The below code will segfault when a virtual thread is yielded. Because `targetThread` will be null for a yielded thread. We need to protect it with `(targetThread == NULL)`.
```
		rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE, &isVirtual); <- Will pin the j.l.VirtualThread
		if ((rc == JVMTI_ERROR_NONE) && (NULL == targetThread)) { <- Protect for null targetThread
			rv_threadid = (jlong) omrthread_get_osId(targetThread->osThread);
			releaseVMThread(currentThread, targetThread, thread, isVirtual); <- Won't be executed for null targetThread so the j.l.VirtualThread won't be unpinned
		}
```

In other JVMTI functions, we will need to stack-allocate a `J9VMThread` if a virtual thread is yielded.

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: > if a user sends a NULL thread (as input) and getVMThread allows a null. in such a case, isVirtual will stay FALSE, and none of the new code will be invoked.

So doesnt this mean you dont need the `isVirtual` parameter? Because if `thread` is NULL then `isVirtual` must be false. And if `thread` is not null, you can always determine if it is virtual.

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: And Im only talking about `releaseVMThread`.

Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c:175]: `BOOLEAN isVirtual` allows us to share the result of `IS_VIRTUAL_THREAD` after evaluating it once in `getVMThread`. `IS_VIRTUAL_THREAD == isSameOrSuperClassOf` will be invoked multiple times if `BOOLEAN isVirtual` is removed.

Comment (tajila) [reply runtime/jvmti/jvmtiHelpers.c:175]: Yes, I understand. My issue is that the `isVirtual` is relevant in `getVMThread` and `releaseVMThread`, but not always in the caller. It makes it confusing to define variables that have no meaning in the context that they are defined in. 

I think it would be cleared for `jvmtiGetOSThreadID` to be

```
	rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE);
		rc = getVMThread(currentThread, thread, &targetThread, TRUE, TRUE, NULL);
		if (rc == JVMTI_ERROR_NONE) {
			rv_threadid = (jlong) omrthread_get_osId(targetThread->osThread);
			releaseVMThread(currentThread, targetThread);
			releaseVMThread(currentThread, targetThread, thread);
```

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

File: runtime/jvmti/jvmtiHelpers.c
Comment (babsingh) [reply runtime/jvmti/jvmtiHelpers.c]: Updated.

**Review by hangshao0 — APPROVED** [review]

**Comment by tajila** [comment]
@gacholio any concerns with this

**Comment by tajila** [comment]
jenkins test sanity amac jdk19

**Comment by tajila** [comment]
jenkins compile win jdk19

**Comment by tajila** [comment]
jenkins test sanity amac jdk19

**Comment by fengxue-IS** [comment]
I've added review comments https://github.com/eclipse-openj9/openj9/pull/15766#discussion_r953142285, I think we need to address them before merging.

File: runtime/jcl/common/thread.cpp
Comment (gacholio) [inline runtime/jcl/common/thread.cpp:538]: Why is the count not checked in this case?

**Comment by gacholio** [comment]
I see a lot of releaseVMThread being called with NULL thread - why is this? Won't it leave the counter unbalanced?

**Comment by gacholio** [comment]
This same code needs to go in `haltThreadForInspection`, I think, so the comments in the general code should not reference JVMTI.

**Comment by hangshao0** [comment]
FYI @gacholio, Eric has taken over this PR and review is being done under https://github.com/eclipse-openj9/openj9/pull/15766

**Comment by gacholio** [comment]
Thanks, I'll look over there.

**Pull request closed without merging**
